kuma/api/v1/plus/bookmarks.py

```diff
@@ -0,0 +1,197 @@
+import functools
+
+
+from django.conf import settings
+from django.http import (
+    HttpResponseBadRequest,
+    HttpResponseForbidden,
+    JsonResponse,
+)
+from django.middleware.csrf import get_token
+from django.utils import timezone
+from django.views.decorators.cache import never_cache
+from django.views.decorators.http import require_http_methods
+
+from kuma.bookmarks.models import Bookmark
+from kuma.documenturls.models import DocumentURL, download_url
+
+
+class NotOKDocumentURLError(Exception):
+    """When the document URL doesn't resolve as a 200 OK"""
+
+
+def require_subscriber(func):
+    @functools.wraps(func)
+    def inner(request, *args, **kwargs):
+        user = request.user
+        if not user.is_authenticated:
+            return HttpResponseForbidden("not signed in")
+        # TEMPORARY until all things auth + subscription come together.
+        if not settings.FAKE_USER_SUBSCRIBER_NUMBER:
+            return HttpResponseForbidden("not a subscriber")
+        # Note: Deliberately commented out until we have figured out OIDC!
+        # if not UserSubscription.objects.filter(
+        #     user=user, canceled__isnull=True
+        # ).exists():
+        #     return HttpResponseForbidden("not a subscriber")
+        return func(request, *args, **kwargs)
+
+    return inner
+
+
+@never_cache
+@require_http_methods(["GET", "POST"])
+@require_subscriber
+def bookmarks(request):
+    # E.g. POST -d url=https://... /api/v1/bookmarks/
+    # or   GET /api/v1/bookmarks/?url=https://...
+    if request.method == "POST" or request.GET.get("url"):
+        return _toggle_bookmark(request)
+
+    # E.g. GET /api/v1/bookmarks/
+    return _get_bookmarks(request)
```

Comment:
I was thinking that we'd always use `_get_bookmarks` for both of the `GET` cases:

1.  When a `url` param is provided, we'd return either an empty list if the subscriber hasn't bookmarked that `url`, or a list of that one bookmark if they have. For this case we'd ignore any `page` and `per_page` params if provided.
2. When a `url` param is **not** provided, we'd respect the `page` and `per_page` params if provided and return a list of all of that subscriber's bookmarks.

but if you'd rather keep it the way it is currently, where the `GET` with a `url` param returns a differently-formatted JSON response than the `GET` without a `url` param, could you move the chunk of code within `_toggle_bookmark` that handles the `GET` case into its own new function called `_get_bookmark` (or something like that)? I think it would make things more readable to see something like this:

```python
    if request.method == "POST":
        return _toggle_bookmark(request)

    if request.GET.get("url"):
        return _get_bookmark(request)

    return _get_bookmarks(request)
```

so readers of the code don't mistakenly think that we're going to toggle the bookmark for `GET`s with a `url` because the function is named `_toggle_bookmark`.

Reply:
> readers of the code don't mistakenly think that we're going to toggle the bookmark for `GET`s

Fair point but "toggle" is a write-operation so GET is never appropriate for that. 
Having said that, it's weird that the function is called `_toggle_bookmark()` when it can take a GET request :)

My reasoning, perhaps flawed, was that finding out if the page you're on is bookmarked or not is possibly going to be very different in nature. I.e. you don't need metadata stuff like:
```python
"metadata": {
    "total": users_bookmarks.count(),
    "page": page,
    "per_page": per_page,
},
```

And what if we decide to need very different stuff from that `items` list if it's *not* an empty list. 

Now, you're making a slightly different suggestion that we break it out into *3* private functions. I.e. `_get_bookmark`. 
The reasoning I lumped what *you* call `_toggle_bookmark` and `_get_bookmark` is that they do a lot of stuff in common. But the more I look at it, perhaps it's sufficiently easy to make that functionality shared across 2 different functions.

Reply:
@peterbe Yeah, my main concern was that the code reads like a `GET` with a `url` param will toggle the bookmark, but of course you see that it doesn't once you look inside the `_toggle_bookmark` function. It's just that the name of the function is misleading for the `GET` case.

Reply:
Implemented now. Thanks for the patience and feedback!